worker/handler.ts

```diff
@@ -95,26 +99,91 @@ export function extractAltLineNumberReference(
   };
 }
 
+interface CacheEntry {
+  body: ArrayBuffer;
+  contentType: string;
+  etag: string;
+  immutable: boolean;
+}
+
+const cache = new Map<string, CacheEntry>();
+
+// When deploying a new version, the worker will be updated before the static
+// website hosted by Vercel gets updated. Therefore we clear the cache 2 minutes
+// after startup up to ensure that there is no stale static content in the
+// in-memory cache.
+setTimeout(() => cache.clear(), 2 * 60 * 1000);
+
 async function proxyFile(
   url: URL,
   remoteUrl: string,
   request: Request,
 ): Promise<Response> {
-  const init = {
-    method: request.method,
-    headers: request.headers,
-  };
-  const urlR = remoteUrl + url.pathname;
-  const modifiedRequest = new Request(urlR, init);
-  let lastErr;
+  const proxyUrl = new URL(remoteUrl + url.pathname).href;
+  let cacheEntry = cache.get(proxyUrl);
+
+  if (cacheEntry === undefined) {
+    const proxyRequest = new Request(proxyUrl);
+    const proxyResponse = await fetchWithRetry(proxyRequest);
+
+    if (!(proxyResponse.ok || proxyResponse.redirected)) {
+      return proxyResponse;
+    }
+
+    const body = await proxyResponse.arrayBuffer();
+    const contentType = proxyResponse.headers.get("content-type") ??
+      "application/binary";
+    const etag = await crypto.subtle.digest("SHA-1", body).then((hash) =>
+      Array.from(new Uint8Array(hash))
+        .map((byte) => byte.toString(16).padStart(2, "0"))
+        .join("")
+    );
+    const immutable = /\bimmutable\b/i
+      .test(proxyResponse.headers.get("cache-control") ?? "");
+
+    cacheEntry = {
+      body,
+      contentType,
+      etag,
+      immutable,
+    };
+    cache.set(proxyUrl, cacheEntry);
+  }
+
+  if (request.headers.get("if-none-match") !== cacheEntry.etag) {
+    let body;
+    switch (request.method) {
+      case "HEAD":
+        body = null;
+        break;
+      case "GET":
+        body = cacheEntry.body;
+        break;
+      default:
+        throw new Error(`Unsupported request method: ${request.method}`);
+    }
+    return new Response(body, {
+      headers: {
+        "content-type": cacheEntry.contentType,
+        "cache-control": cacheEntry.immutable ? "public,immutable" : "private",
+        "etag": cacheEntry.etag,
+      },
```

Comment:
Why only these 3 headers?

Reply:
Because they're the only ones that matter.

Note that I do not forward any request headers to Vercel either, because they might (in theory, not in practice) change the response, which would make the cache too simplistic.

Which ones would you like to see included?

Reply:
Seems fine.